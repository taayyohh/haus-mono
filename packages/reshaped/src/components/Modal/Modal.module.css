.root {
  transition: transform var(--rs-easing-accelerate) var(--rs-duration-medium);
  background: var(--rs-color-background-elevation-overlay);
  color: var(--rs-color-foreground-neutral);
  box-shadow: var(--rs-shadow-overlay);
}

@responsive .root {
  @variable --rs-modal-size calc(var(--rs-unit-x1) * 100);
}

@responsive .--position {
  @value center {
    --rs-modal-size-s: calc(var(--rs-unit-x1) * 100);

    width: var(--rs-modal-size);
    height: auto;
    max-width: calc(100vw - var(--rs-unit-x8));
    max-height: none;
    margin: var(--rs-unit-x4);
    inset: 0;
    transform: scale(0.96);
    position: relative;
    border-radius: var(--rs-unit-radius-large);
    overflow: hidden;
  }

  @value bottom {
    --rs-modal-size-s: auto;

    width: 100%;
    height: var(--rs-modal-size);
    max-width: 100%;
    max-height: calc(100vh - var(--rs-unit-x4));
    margin: 0;
    margin-top: var(--rs-unit-x4);
    inset: 0;
    inset-block-start: initial;
    transform: translate(0, 100%);
    position: fixed;
    border-radius: var(--rs-unit-radius-large) var(--rs-unit-radius-large) 0 0;
    overflow: auto;
  }

  @value start {
    --rs-modal-size-s: calc(var(--rs-unit-x1) * 100);

    width: var(--rs-modal-size);
    height: 100%;
    max-width: calc(100vw - var(--rs-unit-x4));
    max-height: 100%;
    margin: 0;
    margin-inline-end: var(--rs-unit-x4);
    inset: 0;
    inset-inline-end: initial;
    transform: translate(-100%, 0);
    position: fixed;
    border-radius: 0;
    overflow: auto;

    [dir='rtl'] & {
      transform: translate(100%, 0);
    }
  }

  @value end {
    --rs-modal-size-s: calc(var(--rs-unit-x1) * 100);

    width: var(--rs-modal-size);
    height: 100%;
    max-width: calc(100vw - var(--rs-unit-x4));
    max-height: 100%;
    margin: 0;
    margin-inline-start: var(--rs-unit-x4);
    inset: 0;
    inset-inline-start: initial;
    transform: translate(100%, 0);
    position: fixed;
    border-radius: 0;
    overflow: auto;

    [dir='rtl'] & {
      transform: translate(-100%, 0);
    }
  }
}

.--active,
[dir='rtl'] .--active {
  transform: translate(0, 0) scale(1) !important;
  transition-timing-function: var(--rs-easing-decelerate);
}
