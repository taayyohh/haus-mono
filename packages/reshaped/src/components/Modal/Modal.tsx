'use client'

import React from 'react'
import { classNames, responsiveVariables, responsiveClassNames } from 'utilities/helpers'
import Text from 'components/Text'
import Overlay from 'components/Overlay'
import useElementId from 'hooks/useElementId'
import type * as T from './Modal.types'
import s from './Modal.module.css'
import getPaddingStyles from 'styles/padding'

const Context = React.createContext<T.Context>({
  id: '',
  titleMounted: false,
  setTitleMounted: () => {},
  subtitleMounted: false,
  setSubtitleMounted: () => {},
})
const useModal = () => React.useContext(Context)

const ModalTitle = (props: T.TitleProps) => {
  const { children } = props
  const { id, setTitleMounted } = useModal()

  React.useEffect(() => {
    setTitleMounted(true)
    return () => setTitleMounted(false)
  }, [setTitleMounted])

  return (
    <Text variant="featured-3" weight="bold" attributes={{ id: `${id}-title` }}>
      {children}
    </Text>
  )
}

const ModalSubtitle = (props: T.SubtitleProps) => {
  const { children } = props
  const { id, setSubtitleMounted } = useModal()

  React.useEffect(() => {
    setSubtitleMounted(true)
    return () => setSubtitleMounted(false)
  }, [setSubtitleMounted])

  return (
    <Text variant="body-3" color="neutral-faded" attributes={{ id: `${id}-subtitle` }}>
      {children}
    </Text>
  )
}

const Modal = (props: T.Props) => {
  const {
    children,
    onClose,
    active,
    size,
    padding = 4,
    position = 'center',
    transparentOverlay,
    className,
    attributes,
  } = props
  const id = useElementId()
  const [titleMounted, setTitleMounted] = React.useState(false)
  const [subtitleMounted, setSubtitleMounted] = React.useState(false)
  const paddingStyles = getPaddingStyles(padding)

  const value = React.useMemo(
    () => ({
      titleMounted,
      setTitleMounted,
      subtitleMounted,
      setSubtitleMounted,
      id,
    }),
    [id, subtitleMounted, titleMounted]
  )

  return (
    <Overlay onClose={onClose} active={active} transparent={transparentOverlay}>
      {({ active }) => {
        const rootClassNames = classNames(
          s.root,
          className,
          paddingStyles?.classNames,
          active && s['--active'],
          responsiveClassNames(s, '--position', position)
        )

        return (
          <Context.Provider value={value}>
            <div
              {...attributes}
              style={{
                ...paddingStyles?.variables,
                ...responsiveVariables('--rs-modal-size', size),
              }}
              aria-labelledby={titleMounted ? `${id}-title` : undefined}
              aria-describedby={subtitleMounted ? `${id}-subtitle` : undefined}
              className={rootClassNames}
              aria-modal="true"
              role="dialog"
            >
              {children}
            </div>
          </Context.Provider>
        )
      }}
    </Overlay>
  )
}

Modal.Title = ModalTitle
Modal.Subtitle = ModalSubtitle
export default Modal
