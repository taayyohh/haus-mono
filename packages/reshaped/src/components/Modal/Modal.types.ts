import React from "react";
import type * as G from "types/global";

export type Context = {
	id: string;
	titleMounted: boolean;
	setTitleMounted: (b: boolean) => void;
	subtitleMounted: boolean;
	setSubtitleMounted: (b: boolean) => void;
};

export type TitleProps = {
	children: React.ReactNode;
};

export type SubtitleProps = {
	children: React.ReactNode;
};

export type Props = {
	children?: React.ReactNode;
	position?: G.Responsive<"center" | "end" | "bottom" | "start">;
	size?: G.Responsive<string>;
	padding?: G.Responsive<number>;
	active?: boolean;
	transparentOverlay?: boolean;
	onClose?: () => void;
	className?: G.ClassName;
	attributes?: G.Attributes<"div", Props>;
};
