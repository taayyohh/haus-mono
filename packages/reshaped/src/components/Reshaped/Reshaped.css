:root {
  --rs-z-index-raised: 5;
  --rs-z-index-flyout: 80;
  --rs-z-index-fixed: 90;
  --rs-z-index-overlay: 100;
  --rs-z-index-notification: 110;
}

[data-rs-color-mode='light'] {
  color-scheme: light;
}

[data-rs-color-mode='dark'] {
  color-scheme: dark;
}

* {
  box-sizing: border-box;
}

body,
h1,
h2,
h3,
h4,
h5,
h6,
p,
ul,
ol,
li,
figure,
figcaption,
blockquote,
dl,
dd {
  margin: 0;
  padding: 0;
}

ul[class],
ol[class] {
  list-style: none;
}

a:not([class]) {
  text-decoration-skip-ink: auto;
}

img {
  max-width: 100%;
  display: block;
}

input,
button,
textarea,
select {
  font: inherit;
}

/* 
Setting background for Windows browsers that have white background by default 
which doesn't work well in dark mode
*/
option {
  background: var(--rs-color-background-elevation-base);
}

label {
  cursor: pointer;
}

@media (prefers-reduced-motion: reduce) {
  * {
    animation-duration: 0.01ms !important;
    animation-iteration-count: 1 !important;
    transition-duration: 0.01ms !important;
    scroll-behavior: auto !important;
  }
}

[data-rs-no-transition] *,
[data-rs-no-transition] *::before,
[data-rs-no-transition] *::after {
  transition: none !important;
}

input::placeholder {
  color: var(--rs-color-neutral-faded);
  opacity: 0.5;
}

html {
  height: 100%;

  /* https://developer.mozilla.org/en-US/docs/Web/CSS/text-size-adjust */
  -webkit-text-size-adjust: 100%;
  -webkit-font-smoothing: antialiased;

  /* https://developer.mozilla.org/en-US/docs/Web/CSS/text-rendering */
  text-rendering: optimizelegibility;

  /* https://developer.mozilla.org/en-US/docs/Web/CSS/touch-action#values */
  touch-action: manipulation;
}

body {
  height: 100%;
  scroll-behavior: smooth;
  text-rendering: optimizespeed;
  color: var(--rs-color-foreground-neutral);
  background: var(--rs-color-background-page);
}

body,
[data-rs-theme] {
  font-size: var(--rs-font-size-body-3);
  line-height: var(--rs-line-height-body-3);
  font-family: var(--rs-font-family-body);
  font-weight: var(--rs-font-weight-regular);
}

textarea {
  resize: vertical;
}

table {
  border-collapse: collapse;
  border-spacing: 0;
}

fieldset {
  border: 0;
  margin: 0;
  padding: 0;
}

a {
  text-decoration: underline;
  color: var(--rs-color-foreground-primary);

  &:hover {
    text-decoration: none;
  }
}
