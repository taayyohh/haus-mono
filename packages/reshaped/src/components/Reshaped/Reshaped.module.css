.root {
	color: inherit;
	height: inherit;
}
