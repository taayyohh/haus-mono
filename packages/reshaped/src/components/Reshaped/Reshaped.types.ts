import type React from "react";
import type { GlobalColorModeProps, ThemeProps } from "components/Theme";
import type { ToastProviderProps } from "components/Toast";
import type * as G from "types/global";

export type Props = {
	children?: React.ReactNode;
	defaultRTL?: boolean;
	defaultColorMode?: GlobalColorModeProps["defaultMode"];
	defaultViewport?: G.Viewport;
	className?: G.ClassName;
	theme?: NonNullable<ThemeProps["name"]>;
	defaultTheme?: NonNullable<ThemeProps["defaultName"]>;
	toastOptions?: ToastProviderProps["options"];
};
