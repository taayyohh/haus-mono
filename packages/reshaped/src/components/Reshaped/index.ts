export { default } from "./Reshaped";
export type { Props as ReshapedProps } from "./Reshaped.types";
