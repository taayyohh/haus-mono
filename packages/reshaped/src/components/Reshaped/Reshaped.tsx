"use client";

import React from "react";
import { classNames } from "utilities/helpers";
import Theme, { GlobalColorMode } from "components/Theme";
import { ToastProvider } from "components/Toast";
import useSingletonKeyboardMode from "hooks/_private/useSingletonKeyboardMode";
import {
	SingletonEnvironmentContext,
	useSingletonRTL,
} from "hooks/_private/useSingletonEnvironment";
import { SingletonHotkeysProvider } from "hooks/_private/useSingletonHotkeys";
import type * as T from "./Reshaped.types";
import "./Reshaped.css";
import s from "./Reshaped.module.css";

const ReshapedInner = (props: T.Props) => {
	const { children, defaultRTL, defaultViewport = "s", toastOptions } = props;
	const rtlState = useSingletonRTL(defaultRTL);

	useSingletonKeyboardMode();

	return (
		<SingletonEnvironmentContext.Provider value={{ rtl: rtlState, defaultViewport }}>
			<SingletonHotkeysProvider>
				<ToastProvider options={toastOptions}>{children}</ToastProvider>
			</SingletonHotkeysProvider>
		</SingletonEnvironmentContext.Provider>
	);
};

const Reshaped = (props: T.Props) => {
	const { theme, defaultTheme = "reshaped", defaultColorMode, className } = props;
	const rootClassNames = classNames(s.root, className);

	return (
		<GlobalColorMode defaultMode={defaultColorMode}>
			<Theme name={theme} defaultName={defaultTheme} className={rootClassNames}>
				<ReshapedInner {...props}>{props.children}</ReshapedInner>
			</Theme>
		</GlobalColorMode>
	);
};

export default Reshaped;
