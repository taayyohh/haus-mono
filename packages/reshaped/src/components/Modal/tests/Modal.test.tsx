import React from "react";
import { vi } from "vitest";
import { render, screen, waitFor, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Modal from "components/Modal";
import Reshaped from "components/Reshaped";

const fixtures = {
	content: "Content",
	title: "Title",
	subtitle: "Subtitle",
	className: "test-className",
	id: "test-id",
};

describe("Components/Modal", () => {
	test("doesn't render children", () => {
		render(
			<Reshaped>
				<Modal>{fixtures.content}</Modal>
			</Reshaped>
		);

		expect(screen.queryByText(fixtures.content)).not.toBeInTheDocument();
	});

	test("renders correctly when active", () => {
		render(
			<Reshaped>
				<Modal active>
					<Modal.Title>{fixtures.title}</Modal.Title>
					<Modal.Subtitle>{fixtures.subtitle}</Modal.Subtitle>
					{fixtures.content}
				</Modal>
			</Reshaped>
		);

		const elModal = screen.getByRole("dialog");
		const elTitle = screen.getByText(fixtures.title);
		const elSubtitle = screen.getByText(fixtures.subtitle);
		const titleId = elTitle.getAttribute("id");
		const subtitleId = elSubtitle.getAttribute("id");

		expect(elModal).toBeInTheDocument();
		expect(elTitle).toBeInTheDocument();
		expect(elSubtitle).toBeInTheDocument();
		expect(titleId).toMatchSnapshot();
		expect(subtitleId).toMatchSnapshot();

		expect(elModal).toHaveAttribute("aria-modal", "true");
		expect(elModal).toHaveAttribute("aria-labelledby", titleId);
		expect(elModal).toHaveAttribute("aria-describedby", subtitleId);
	});

	test("works with className and attributes", () => {
		render(
			<Reshaped>
				<Modal active className={fixtures.className} attributes={{ id: fixtures.id }}>
					{fixtures.content}
				</Modal>
			</Reshaped>
		);

		const elModal = screen.getByRole("dialog");

		expect(elModal).toHaveClass(fixtures.className);
		expect(elModal).toHaveAttribute("id", fixtures.id);
	});

	test("works as controlled", async () => {
		const handleCloseMock = vi.fn();

		const Component = () => {
			const [active, setActive] = React.useState(true);

			const handleOpen = () => {
				setActive(true);
			};

			const handleClose = () => {
				setActive(false);
				handleCloseMock();
			};

			return (
				<Reshaped>
					<button type="button" onClick={handleOpen}>
						Open
					</button>
					<Modal active={active} onClose={handleClose}>
						{fixtures.content}
					</Modal>
				</Reshaped>
			);
		};
		render(<Component />);

		expect(screen.queryByText(fixtures.content)).toBeInTheDocument();

		await userEvent.keyboard("{Escape}");

		/**
		 * Need to trigger transition event on the correct element
		 * since Overlay ignores event bubbling to prevent it from closing based on other elements transitionend
		 */
		const overlayEl = screen.getAllByRole("button")[1];
		fireEvent.transitionEnd(overlayEl, {
			propertyName: "opacity",
		});

		await waitFor(() => {
			expect(handleCloseMock).toBeCalledTimes(1);
			expect(screen.queryByText(fixtures.content)).not.toBeInTheDocument();
		});

		const elButton = screen.getByText("Open");

		await userEvent.click(elButton);

		expect(screen.getByText(fixtures.content)).toBeInTheDocument();
	});
});
