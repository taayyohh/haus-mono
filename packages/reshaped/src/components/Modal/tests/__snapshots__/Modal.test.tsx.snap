// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Components/Modal > renders correctly when active 1`] = `":r2:-title"`;

exports[`Components/Modal > renders correctly when active 2`] = `":r2:-subtitle"`;
